### ISSUE TITLE
Should @local_config_platform//:host be marked no-remote-exec (or maybe even no-remote)?

### ISSUE THREAD

**Issue by EdSchouten** [bazelbuild/bazel#22777]
### Description of the feature request:

As of #22363, we have testing coverage in place for ensuring platforms as a whole can be marked no-remote-exec. This raises the question: should the default host platform be marked as such?

Think of it like this: the default host platform does not have any remote execution platform properties set up. This means that if such a platform were to be used for remote execution, it would likely just result in an error indicating that no workers exist having the empty set of platform properties. By marking it as no-remote-exec, we're essentially forcing the user to declare its own platform that has the right properties set.

### What underlying problem are you trying to solve with this feature?

I've got a simple remote execution cluster that only has Linux workers, which I want to use from my Mac to speed up builds where possible. However, there are some actions that only run on macOS, which I want to run locally. So I declare two platforms:

```python
platform(
    name = "remote_linux",
    constraint_values = [
        "@platforms//cpu:x86_64",
        "@platforms//os:linux",
    ],
    exec_properties = { ... },
)

platform(
    name = "local_fallback",
    exec_properties = {"no-remote-exec": "true"},
    parents = ["@local_config_platform//:host"],
)
```

And then call Bazel with `--extra_execution_platforms=//:remote_linux,//:local_fallback`. If `@local_config_platform//:host` itself were marked `no-remote-exec`, I could get rid of that second platform declaration.

### Which operating system are you running Bazel on?

macOS

### What is the output of `bazel info release`?

7.x

**Comment by sluongng** [comment]
Given the current observed behavior in https://github.com/bazelbuild/bazel/pull/23037, and several related questions/discussions around this topic in:

- https://github.com/bazelbuild/bazel/issues/10799#issuecomment-2246634317
- https://github.com/bazelbuild/bazel/pull/17390
- https://github.com/bazelbuild/bazel/discussions/20497
- https://bazelbuild.slack.com/archives/C01E7TH8XK9/p1721781699224499

We should hold off on promoting this usage pattern until we figure out how to address the test's execution platform. On our end, we are seeing remote test actions sent to our RBE server with `"no-remote-exec": "true"` which is very confusing to our users.

---

The current usage pattern also a bit overlap with dynamic execution. One could set up the build so that the action is executed both remotely and locally, where remote exec points to a non-existent pool of workers and never gets executed, forcing it to be run locally instead.

---

Finally, semi-related but I hope we can get a bit more telemetry data about platforms and toolchains via https://github.com/bazelbuild/bazel/pull/23019. It would help us build tools to help detect badly configured platforms/toolchains.

**Comment by moroten** [comment]
Is it time to revisit `no-remote-exec` for the host platform now when #25160 is done?

**Comment by fmeum** [comment]
Yes, I think it's time to consider that. The change would need to be made in `@platforms`, which raises some questions about how to control this breaking change. We could artificially tie it to a Bazel version or gate it behind a flag forever.

**Comment by moroten** [comment]
What about `--host_platform=@platforms//host:host_with_remote_execution` for the old behaviour?

**Comment by sluongng** [comment]
i think this is probably a good time to revisit this and aim for it to be an incompatible change in the next major release (10). We just need to make sure that it behaves as expected for folks who were using the current default platform for RBE (with exec properties set via flags). We would also need to assess how it would work with dynamic execution use cases.

@moroten do you want to work on this?

**Comment by moroten** [comment]
> We would also need to assess how it would work with dynamic execution use cases.

It sounds like the registered execution platforms should ideally inform if they are meant for execution locally, remote or both. Maybe the pattern `exec_properties = {"no-remote-exec": "true"}` and `exec_properties = {"no-local-exec": "true"}` is a bit clumsy. Should it be `register_execution_platforms(location, platforms)` where the default is `location = ["remote", "local"]`?

> [@moroten](https://github.com/moroten) do you want to work on this?

I can give it a try in a few weeks.
